File: src/main/java/org/noxet/noxetserver/playerstate/properties/PSPExperienceProgress.java
package org.noxet.noxetserver.playerstate.properties;

import org.bukkit.entity.Player;
import org.noxet.noxetserver.playerstate.PlayerStateProperty;

public class PSPExperienceProgress implements PlayerStateProperty<Float> {
    @Override
    public String getConfigName() {
        return "experience_progress";
    }

    @Override
    public Float getDefaultSerializedProperty() {
        return 0F;
    }

    @Override
    public Float getSerializedPropertyFromPlayer(Player player) {
        return player.getExp();
    }

    @Override
    public void restoreProperty(Player player, Float progress) {
        player.setExp(progress);
    }

    @Override
    public Class<Float> getTypeClass() {
        return Float.class;
    }
}

File: src/main/java/org/noxet/noxetserver/playerstate/properties/PSPExperienceProgress.scala
package org.noxet.noxetserver.playerstate.properties

import org.bukkit.configuration.ConfigurationSection
import org.bukkit.entity.Player
import org.noxet.noxetserver.playerstate.PlayerStateProperty

object PSPExperienceProgress extends PlayerStateProperty[Float]:
  override def getConfigName: String = "experience_progress"

  override def getDefaultSerializedProperty: Float = 0F

  override def getSerializedPropertyFromPlayer(player: Player): Float =
    player.getExp

  override def restoreProperty(player: Player, progress: Float): Unit =
    player.setExp(progress)

  override def getValueFromConfig(config: ConfigurationSection): Float =
    config.getDouble(getConfigName).toFloat